src/app-layer-ssl.c

```diff
@@ -258,55 +263,214 @@ static int SSLSetTxMpmIDs(void *vtx, uint64_t mpm_ids)
     }
 
-    input += session_id_length;
+    *input += session_id_length;
 
+    return 0;
+}
+
+static inline int TLSDecodeHSHelloCipherSuites(SSLState *ssl_state,
+                                               uint8_t **input,
+                                               const uint32_t input_len,
+                                               const uint8_t *initial_input)
+{
     if (!(HAS_SPACE(2)))
         goto invalid_length;
 
-    /* skip cipher suites */
-    uint16_t cipher_suites_length = input[0] << 8 | input[1];
-    input += 2;
-
-    input += cipher_suites_length;
+    uint16_t cipher_suites_length = **input << 8 | *(*input + 1);
+    *input += 2;
 
-    if (!(HAS_SPACE(1)))
+    if (!(HAS_SPACE(cipher_suites_length)))
         goto invalid_length;
 
-    /* skip compression methods */
-    uint8_t compression_methods_length = *(input++);
+    if ((ssl_state->current_flags & SSL_AL_FLAG_STATE_CLIENT_HELLO) &&
+            ssl_config.enable_ja3) {
+        int rc;
+
+        JA3Buffer *ja3_cipher_suites = Ja3BufferInit();
+        if (ja3_cipher_suites == NULL)
+            return -1;
+
+        uint16_t processed_len = 0;
+        while (processed_len < cipher_suites_length)
+        {
+            if (!(HAS_SPACE(2)))
+                goto invalid_length;
+
+            uint16_t cipher_suite = **input << 8 | *(*input + 1);
+            *input += 2;
+
+            if (TLSDecodeValueIsGREASE(cipher_suite) != 1) {
+                rc = Ja3BufferAddValue(ja3_cipher_suites, cipher_suite);
+                if (rc != 0) {
+                    Ja3BufferFree(ja3_cipher_suites);
+                    return -1;
+                }
+            }
+
+            processed_len += 2;
+        }
+
+        rc = Ja3BufferAppendBuffer(ssl_state->ja3_str, ja3_cipher_suites);
+        if (rc != 0) {
+            Ja3BufferFree(ja3_cipher_suites);
+            return -1;
+        }
+
+    } else {
+        /* Skip cipher suites */
+        *input += cipher_suites_length;
+    }
 
-    input += compression_methods_length;
+    return 0;
+
+invalid_length:
+    SCLogDebug("TLS handshake invalid length");
+    SSLSetEvent(ssl_state,
+                TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+    return -1;
+}
 
-    /* extensions are optional (RFC5246 section 7.4.1.2) */
+static inline int TLSDecodeHSHelloCompressionMethods(SSLState *ssl_state,
+                                                     uint8_t **input,
+                                                     const uint32_t input_len,
+                                                     const uint8_t *initial_input)
+{
+    if (!(HAS_SPACE(1))) {
+        SCLogDebug("TLS handshake invalid length");
+        SSLSetEvent(ssl_state,
+                    TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+        return -1;
+    }
+
+    /* Skip compression methods */
+    uint8_t compression_methods_length = **input;
+    *input += 1;
+
+    *input += compression_methods_length;
```

Comment:
should we first have a check for HAS_SPACE(compression_methods_length)?

Reply:
Do we need it? We are not dereferencing input as far as I can see.

Reply:
it's moved forward for the next caller isn't it? Seems unsafe to do this w/o check?

Reply:
Yes, it is. I'll add a check if that makes you sleep better at night ;)